File: core/planner.py
"""이동 계획 수립: 전략 적용 → 충돌 해결 → 흐름(출발 폴더 → 도착 폴더) 집계."""
from __future__ import annotations

import os
from collections import defaultdict

from .scanner import FileInfo
from .strategies import Strategy, category_of
from . import i18n


def _unique_dest(dest: str, taken: set[str]) -> str:
    key = os.path.normcase(dest)
    if key not in taken and not os.path.exists(dest):
        taken.add(key)
        return dest
    base, ext = os.path.splitext(dest)
    i = 1
    while True:
        cand = f"{base} ({i}){ext}"
        k = os.path.normcase(cand)
        if k not in taken and not os.path.exists(cand):
            taken.add(k)
            return cand
        i += 1


def _first_rule(rules: list[dict], f: FileInfo) -> dict | None:
    from . import rules as _rl
    return _rl.first_match(rules, f)


def build_plan(files: list[FileInfo], root: str, strategy: Strategy, dest_root: str | None = None,
               include_warn: bool = False, exclude_exts: set[str] | None = None,
               min_size: int = 0, max_moves: int | None = None, rules: list[dict] | None = None) -> dict:
    root = os.path.abspath(root)
    dest_root = os.path.abspath(dest_root or root)
    exclude_exts = {e.lower().lstrip(".") for e in (exclude_exts or set())}

    candidates = [
        f for f in files
        if f.risk == "safe" or (include_warn and f.risk == "warn")
    ]
    candidates = [f for f in candidates if f.ext not in exclude_exts and f.size >= min_size]
    strategy.prepare(candidates, root)

    moves: list[dict] = []
    taken: set[str] = set()
    skipped_same = 0
    flows: dict[tuple[str, str], dict] = defaultdict(lambda: {"count": 0, "size": 0})
    dest_dirs: dict[str, dict] = defaultdict(lambda: {"count": 0, "size": 0})
    src_dirs: dict[str, dict] = defaultdict(lambda: {"count": 0, "size": 0})

    for f in candidates:
        # 규칙(Pro)이 전략보다 먼저. 처음 맞는 하나만 적용된다.
        rel = None
        rule_name = None
        file_name = f.name
        matched = _first_rule(rules, f) if rules else None
        if matched is not None:
            from . import rules as _rl
            rule_name = matched.get("name") or matched.get("id")
            act = matched.get("action", "folder")
            if act == "skip":
                continue
            if act == "folder":
                exp = _rl.expand(matched.get("value", ""), f).strip()   # 폴더 이름에서만 공백 정리
                if exp:
                    rel = exp
            elif act == "prefix":
                file_name = _rl.expand(matched.get("value", ""), f) + f.name
            elif act == "suffix":
                stem, ext_ = os.path.splitext(f.name)
                file_name = stem + _rl.expand(matched.get("value", ""), f) + ext_
        if rel is None:
            rel = strategy.dest_dir(f, root)
        if rel is None:
            continue
        rel = os.path.normpath(rel)
        target_dir = os.path.join(dest_root, rel)
        if os.path.normcase(target_dir) == os.path.normcase(os.path.dirname(f.path)):
            skipped_same += 1
            continue
        # 목적지 폴더 안에 이미 있는 파일을 다시 그 안으로 넣는 경우 방지 (하위 정리 시)
        dest = _unique_dest(os.path.join(target_dir, file_name), taken)
        src_rel = os.path.relpath(os.path.dirname(f.path), root)
        src_rel = i18n.name("root") if src_rel == "." else src_rel.split(os.sep)[0]
        top_dest = rel.split(os.sep)[0]
        _k, label = category_of(f)
        moves.append({
            "src": f.path, "dst": dest, "name": f.name, "size": f.size, "ext": f.ext,
            "risk": f.risk, "reasons": f.reasons, "src_group": src_rel, "dst_group": top_dest,
            "dst_rel": rel, "category": label, "renamed": os.path.basename(dest) != f.name,
            "note": strategy.note(f), "rule": rule_name,
        })
        fl = flows[(src_rel, top_dest)]
        fl["count"] += 1; fl["size"] += f.size
        dest_dirs[rel]["count"] += 1; dest_dirs[rel]["size"] += f.size
        src_dirs[src_rel]["count"] += 1; src_dirs[src_rel]["size"] += f.size
        if max_moves and len(moves) >= max_moves:
            break

    total_size = sum(m["size"] for m in moves)
    warn_count = sum(1 for m in moves if m["risk"] == "warn")
    renamed = sum(1 for m in moves if m["renamed"])
    return {
        "root": root, "dest_root": dest_root, "strategy": strategy.id,
        "moves": moves,
        "summary": {
            "move_count": len(moves), "total_size": total_size, "warn_count": warn_count,
            "renamed": renamed, "skipped_same": skipped_same, "candidates": len(candidates),
        },
        "flows": [{"src": s, "dst": d, **v} for (s, d), v in sorted(flows.items(), key=lambda kv: -kv[1]["size"])],
        "dest_tree": _tree(dest_dirs),
        "src_groups": [{"name": k, **v} for k, v in sorted(src_dirs.items(), key=lambda kv: -kv[1]["size"])],
    }


def flows_at(moves: list[dict], dest: str = "") -> dict:
    """흐름도를 한 단계 안으로 들어가서 본다. dest 가 빈 문자열이면 최상위. (C# Planner.Flows 와 동일)"""
    sep = os.sep
    dest = (dest or "").strip().replace("/", sep).strip(sep)   # 화면은 / 를 쓸 수 있다
    prefix = dest + sep if dest else ""
    if dest:
        scoped = [m for m in moves
                  if m["dst_rel"].lower() == dest.lower() or m["dst_rel"].lower().startswith(prefix.lower())]
    else:
        scoped = moves

    def rest(m):
        r = m["dst_rel"] if not dest else (m["dst_rel"][len(prefix):] if len(m["dst_rel"]) > len(prefix) else "")
        return r.split(sep) if r else []

    flows: dict[tuple[str, str], dict] = defaultdict(lambda: {"count": 0, "size": 0})
    src_dirs: dict[str, dict] = defaultdict(lambda: {"count": 0, "size": 0})
    nodes: dict[str, dict] = defaultdict(lambda: {"count": 0, "size": 0})
    deeper: set[str] = set()

    for m in scoped:
        parts = rest(m)
        seg = parts[0] if parts else ""
        if len(parts) > 1:
            deeper.add(seg)
        for d, k in ((flows, (m["src_group"], seg)), (src_dirs, m["src_group"]), (nodes, seg)):
            d[k]["count"] += 1
            d[k]["size"] += m["size"]

    return {
        "path": dest,
        "crumbs": dest.split(sep) if dest else [],
        "move_count": len(scoped),
        "total_size": sum(m["size"] for m in scoped),
        "flows": [{"src": s, "dst": d, **v} for (s, d), v in sorted(flows.items(), key=lambda kv: -kv[1]["size"])],
        "src_groups": [{"name": k, **v} for k, v in sorted(src_dirs.items(), key=lambda kv: -kv[1]["size"])],
        "dest_nodes": [{"name": k, **v, "has_children": k in deeper} for k, v in sorted(nodes.items(), key=lambda kv: -kv[1]["size"])],
    }


def _tree(dest_dirs: dict[str, dict]) -> list[dict]:
    root: dict = {}
    for rel, v in dest_dirs.items():
        node = root
        for part in rel.split(os.sep):
            node = node.setdefault(part, {"_count": 0, "_size": 0, "_children": {}})
            node["_count"] += v["count"]; node["_size"] += v["size"]
            node = node["_children"]

    def conv(d: dict) -> list[dict]:
        out = []
        for name, n in sorted(d.items(), key=lambda kv: -kv[1]["_size"]):
            out.append({"name": name, "count": n["_count"], "size": n["_size"], "children": conv(n["_children"])})
        return out
    return conv(root)

File: FoldeClean/Core/Planner.cs
using System.IO;

namespace FoldeClean;

public sealed class Plan
{
    public string Root = "", DestRoot = "", StrategyId = "";
    public List<MoveRec> Moves = new();
    public Dictionary<string, object?> Summary = new();
    public List<Dictionary<string, object?>> Flows = new();
    public List<Dictionary<string, object?>> DestTree = new();
    public List<Dictionary<string, object?>> SrcGroups = new();

    public Dictionary<string, object?> ToDict(int maxMoves = 5000) => new()
    {
        ["root"] = Root, ["dest_root"] = DestRoot, ["strategy"] = StrategyId,
        ["moves"] = Moves.Take(maxMoves).ToList(), ["moves_total"] = Moves.Count,
        ["summary"] = Summary, ["flows"] = Flows, ["dest_tree"] = DestTree, ["src_groups"] = SrcGroups,
    };
}

/// <summary>이동 계획: 전략 적용 → 이름 충돌 해결 → 흐름/트리 집계.</summary>
public static class Planner
{
    sealed class Agg { public int Count; public long Size; }

    static string UniqueDest(string dest, HashSet<string> taken)
    {
        if (taken.Add(dest.ToLowerInvariant()) && !File.Exists(dest)) return dest;
        var dir = Path.GetDirectoryName(dest) ?? ""; var stem = Path.GetFileNameWithoutExtension(dest); var ext = Path.GetExtension(dest);
        for (int i = 1; ; i++)
        {
            var cand = Path.Combine(dir, $"{stem} ({i}){ext}");
            if (!taken.Contains(cand.ToLowerInvariant()) && !File.Exists(cand)) { taken.Add(cand.ToLowerInvariant()); return cand; }
        }
    }

    /// <summary>
    /// 이동 계획을 세운다. rules 가 주어지면(Pro) 전략보다 먼저 평가하고, 처음 맞는 규칙 하나만 적용한다.
    /// 규칙 동작: folder(목적지 지정) · skip(옮기지 않음) · prefix/suffix(이름 앞뒤에 붙이고 전략 폴더는 그대로)
    /// </summary>
    public static Plan Build(List<FileRec> files, string root, Strategy strategy, bool includeWarn, HashSet<string> excludeExts, long minSize,
                             Action<string, long, long>? progress = null, RuleSet? rules = null)
    {
        root = Path.GetFullPath(root);
        var plan = new Plan { Root = root, DestRoot = root, StrategyId = strategy.Id };
        var candidates = files.Where(f => (f.Risk == "safe" || (includeWarn && f.Risk == "warn")) && !excludeExts.Contains(f.Ext) && f.Size >= minSize).ToList();
        strategy.Prepare(candidates, root, progress);
        progress?.Invoke("plan", 0, candidates.Count);
        int processed = 0;

        var taken = new HashSet<string>();
        int skippedSame = 0, ruleHits = 0, ruleSkips = 0;
        var flows = new Dictionary<(string, string), Agg>();
        var destDirs = new Dictionary<string, Agg>();
        var srcDirs = new Dictionary<string, Agg>();

        foreach (var f in candidates)
        {
            if (++processed % 500 == 0) progress?.Invoke("plan", processed, candidates.Count);

            // 1) 규칙 (Pro) — 처음 맞는 하나만
            string? rel = null, ruleName = null, fileName = f.Name;
            var matched = rules?.FirstMatch(f, root);
            if (matched != null)
            {
                ruleName = string.IsNullOrWhiteSpace(matched.Name) ? matched.Id : matched.Name;
                switch (matched.Action)
                {
                    case "skip":
                        ruleSkips++; continue;
                    case "folder":
                        var expanded = RuleSet.Expand(matched.Value, f).Trim();   // 폴더 이름에서만 공백 정리
                        if (expanded.Length > 0) rel = expanded;
                        break;
                    case "prefix":
                        fileName = RuleSet.Expand(matched.Value, f) + f.Name; break;
                    case "suffix":
                        var stem = Path.GetFileNameWithoutExtension(f.Name);
                        fileName = stem + RuleSet.Expand(matched.Value, f) + Path.GetExtension(f.Name); break;
                }
                ruleHits++;
            }

            // 2) 규칙이 목적지를 정하지 않았으면 전략에 맡긴다
            rel ??= strategy.DestDir(f, root);
            if (rel == null) continue;
            rel = rel.TrimEnd(Path.DirectorySeparatorChar);
            var targetDir = Path.GetFullPath(Path.Combine(root, rel));
            var srcDir = Path.GetDirectoryName(f.Path) ?? "";
            if (string.Equals(targetDir, srcDir, StringComparison.OrdinalIgnoreCase) && fileName == f.Name) { skippedSame++; continue; }
            var dest = UniqueDest(Path.Combine(targetDir, fileName), taken);
            var srcRel = Path.GetRelativePath(root, srcDir);
            srcRel = srcRel == "." ? I18n.Name("root") : srcRel.Split(Path.DirectorySeparatorChar)[0];
            var topDest = rel.Split(Path.DirectorySeparatorChar)[0];
            plan.Moves.Add(new MoveRec
            {
                Src = f.Path, Dst = dest, Name = f.Name, Size = f.Size, Ext = f.Ext, Risk = f.Risk, Reasons = f.Reasons,
                SrcGroup = srcRel, DstGroup = topDest, DstRel = rel, Category = Categories.Of(f).label,
                Renamed = !string.Equals(Path.GetFileName(dest), f.Name, StringComparison.Ordinal),
                Note = strategy.Note(f), Rule = ruleName,
            });
            Bump(flows, (srcRel, topDest), f.Size); Bump(destDirs, rel, f.Size); Bump(srcDirs, srcRel, f.Size);
        }

        plan.Summary = new()
        {
            ["move_count"] = plan.Moves.Count, ["total_size"] = plan.Moves.Sum(m => m.Size),
            ["warn_count"] = plan.Moves.Count(m => m.Risk == "warn"), ["renamed"] = plan.Moves.Count(m => m.Renamed),
            ["skipped_same"] = skippedSame, ["candidates"] = candidates.Count,
            ["rule_hits"] = ruleHits, ["rule_skips"] = ruleSkips,
        };
        plan.Flows = flows.OrderByDescending(kv => kv.Value.Size).Select(kv => new Dictionary<string, object?> { ["src"] = kv.Key.Item1, ["dst"] = kv.Key.Item2, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size }).ToList();
        plan.SrcGroups = srcDirs.OrderByDescending(kv => kv.Value.Size).Select(kv => new Dictionary<string, object?> { ["name"] = kv.Key, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size }).ToList();
        plan.DestTree = Tree(destDirs);
        return plan;
    }

    /// <summary>
    /// 흐름도를 한 단계 안으로 들어가서 본다. dest 가 빈 문자열이면 최상위.
    /// 오른쪽 칸은 dest 바로 아래 한 단계만 보여주고, 더 들어갈 수 있는지(has_children)도 알려 준다.
    /// dest 폴더에 바로 들어가는 파일은 이름이 "" 인 항목으로 모은다.
    /// </summary>
    public static Dictionary<string, object?> Flows(List<MoveRec> moves, string dest)
    {
        var sep = Path.DirectorySeparatorChar;
        dest = (dest ?? "").Trim().Replace('/', sep).Trim(sep);   // 화면은 / 를 쓸 수 있다
        var prefix = dest.Length == 0 ? "" : dest + sep;
        var scoped = dest.Length == 0 ? moves
            : moves.Where(m => m.DstRel.Equals(dest, StringComparison.OrdinalIgnoreCase)
                            || m.DstRel.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)).ToList();

        // dest 아래로 남은 경로를 조각으로 나눈다
        string[] Rest(MoveRec m)
        {
            var rest = dest.Length == 0 ? m.DstRel
                     : (m.DstRel.Length > prefix.Length ? m.DstRel[prefix.Length..] : "");
            return rest.Length == 0 ? Array.Empty<string>() : rest.Split(sep);
        }

        var flows = new Dictionary<(string, string), Agg>();
        var srcDirs = new Dictionary<string, Agg>();
        var nodes = new Dictionary<string, Agg>();
        var deeper = new HashSet<string>();     // 더 들어갈 수 있는 노드

        foreach (var m in scoped)
        {
            var parts = Rest(m);
            var seg = parts.Length == 0 ? "" : parts[0];
            if (parts.Length > 1) deeper.Add(seg);
            Bump(flows, (m.SrcGroup, seg), m.Size);
            Bump(srcDirs, m.SrcGroup, m.Size);
            Bump(nodes, seg, m.Size);
        }

        return new()
        {
            ["path"] = dest,
            ["crumbs"] = dest.Length == 0 ? new List<string>() : dest.Split(sep).ToList(),
            ["move_count"] = scoped.Count,
            ["total_size"] = scoped.Sum(m => m.Size),
            ["flows"] = flows.OrderByDescending(kv => kv.Value.Size)
                .Select(kv => new Dictionary<string, object?> { ["src"] = kv.Key.Item1, ["dst"] = kv.Key.Item2, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size }).ToList(),
            ["src_groups"] = srcDirs.OrderByDescending(kv => kv.Value.Size)
                .Select(kv => new Dictionary<string, object?> { ["name"] = kv.Key, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size }).ToList(),
            ["dest_nodes"] = nodes.OrderByDescending(kv => kv.Value.Size)
                .Select(kv => new Dictionary<string, object?> { ["name"] = kv.Key, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size, ["has_children"] = deeper.Contains(kv.Key) }).ToList(),
        };
    }

    static void Bump<K>(Dictionary<K, Agg> d, K key, long size) where K : notnull
    {
        if (!d.TryGetValue(key, out var a)) d[key] = a = new Agg();
        a.Count++; a.Size += size;
    }

    sealed class Node { public int Count; public long Size; public Dictionary<string, Node> Children = new(); }

    static List<Dictionary<string, object?>> Tree(Dictionary<string, Agg> destDirs)
    {
        var root = new Dictionary<string, Node>();
        foreach (var (rel, v) in destDirs)
        {
            var level = root;
            foreach (var part in rel.Split(Path.DirectorySeparatorChar))
            {
                if (!level.TryGetValue(part, out var n)) level[part] = n = new Node();
                n.Count += v.Count; n.Size += v.Size;
                level = n.Children;
            }
        }
        List<Dictionary<string, object?>> Conv(Dictionary<string, Node> d) =>
            d.OrderByDescending(kv => kv.Value.Size).Select(kv => new Dictionary<string, object?>
            { ["name"] = kv.Key, ["count"] = kv.Value.Count, ["size"] = kv.Value.Size, ["children"] = Conv(kv.Value.Children) }).ToList();
        return Conv(root);
    }
}
